onnxruntime/core/optimizer/graph_transformer_mgr.h

```diff
@@ -5,31 +5,58 @@
 
 #include "core/optimizer/graph_transformer.h"
 #include "core/optimizer/constant_folding.h"
+#include "core/optimizer/rewrite_rule.h"
+using namespace ::onnxruntime::common;
 
 namespace onnxruntime {
 // Manages a list of graph transformers. It is initialized with a list of graph
 // transformers. Each inference session can further register additional ones.
 class GraphTransformerManager {
  public:
-  explicit GraphTransformerManager(unsigned steps) noexcept : steps_(steps) {
-    // Register default transformers.
-  }
+  explicit GraphTransformerManager(unsigned steps) : steps_(steps) {
+    
+  }  
 
-  // Register a graph transformer.
-  common::Status Register(std::unique_ptr<GraphTransformer> transformer) {
-    transformers_.push_back(std::move(transformer));
-    return common::Status::OK();
-  }
+  // Register a transformer with a level and compatible providers list
+  // If a transformer is execution provider independent add empty string in providers {""}
+  common::Status Register(std::unique_ptr<GraphTransformer> transformer, const TransformerLevel& level, std::vector<std::string>&& provider);  
```

Comment:
Would passing/defaulting to an empty vector be easier? Why is an empty string required?

Reply:
Scenario 1: InsertCast transformer - this is a truly global transformer which is required for correctness
Scenario 2: TransposeElimination, IdentityElimination - these transformers will be applied twice. Once before partitioning and once after partitioning. After partitioning they will only be applied to CpuExecutionProvider nodes... 
for cases like 2 we need a way to say this transformer is applicable in both cases when run before partitioning and when run after partitioning (for Cpu and may be tomorrow for Cuda since that is also an in house provider)

Scenario 1 transformers which are required for correctness will not be managed by transformer manager and there there is no point in defaulting to {} however this makes sense for Transformer->Apply and therefore I did take your suggestion and changed it to default to {}

---
In reply to: [265371635](https://github.com/Microsoft/onnxruntime/pull/608#discussion_r265371635) [](ancestors = 265371635)